File: python-aoc/year2015/day01.py
def solve_part_1(input: str) -> int:
    floor = 0
    for c in input:
        match c:
            case '(': floor += 1
            case ')': floor -= 1
    return floor

def solve_part_2(input: str) -> int:
    floor = 0
    position = 1
    for c in input:
        match c:
            case '(': floor += 1
            case ')': floor -= 1
        if floor < 0:
            return position
        position += 1
    return -1 # unreachable on valid input

File: rust-aoc/src/year2015/day01.rs
use std::ops::Add;

pub fn solve_part_1(input: String) -> i32 {
	input
		.chars()
		.map(|c| match c {
			'(' => 1,
			')' => -1,
			_   => unreachable!(),
		})
		.sum()
}

pub fn solve_part_2(input: String) -> usize {
	input
		.chars()
		.map(|c| match c {
			'(' => 1,
			')' => -1,
			_   => unreachable!(),
		})
		.scan(0, |floor, n| { *floor += n; Some(*floor) })
		.position(|floor| floor < 0)
		.unwrap()
		.add(1)
}